From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: NotAlexNoyle <alexnoyle@icloud.com>
Date: Wed, 19 Nov 2025 12:35:55 -0500
Subject: [PATCH] Prevent server from overwriting existing pufferfish.yml


diff --git a/src/main/java/gg/pufferfish/pufferfish/PufferfishConfig.java b/src/main/java/gg/pufferfish/pufferfish/PufferfishConfig.java
index b4e5fbace85c67e7bd347e6a90514bbc2c132d5e..0d810aa423132d410dd966d9529e6d933400d998 100644
--- a/src/main/java/gg/pufferfish/pufferfish/PufferfishConfig.java
+++ b/src/main/java/gg/pufferfish/pufferfish/PufferfishConfig.java
@@ -53,7 +53,8 @@ public class PufferfishConfig {
 	public static void load() throws IOException {
 		File configFile = pufferfishFile; // Purpur
 		
-		if (configFile.exists()) {
+		boolean newlyCreated = !configFile.exists();
+		if (!newlyCreated) {
 			try {
 				config.load(configFile);
 			} catch (InvalidConfigurationException e) {
@@ -82,7 +83,9 @@ public class PufferfishConfig {
 		
 		updates++;
 		
-		config.save(configFile);
+		if (newlyCreated) {
+			config.save(configFile);
+		}
 		
 		// Attempt to detect vectorization
 		try {
